src/System.Memory/src/System/SpanSortHelpers.Keys.cs

```diff
@@ -0,0 +1,234 @@
+                }
+            }
+        }
+
+        internal interface ISpanSortHelper<TKey>
+        {
+            void Sort(ref TKey keys, int length);
+            void Sort(ref TKey keys, int length, Comparison<TKey> comparison);
+        }
+
+        internal class SpanSortHelper<TKey> : ISpanSortHelper<TKey>
+        {
+            public void Sort(ref TKey keys, int length)
+            {
+                SC.Sort(ref keys, length, Comparer<TKey>.Default);
+            }
+
+            public void Sort(ref TKey keys, int length, Comparison<TKey> comparison)
+            {
+                S.Sort(ref keys, length, comparison);
+            }
+        }
+
+        internal class ComparableSpanSortHelper<TKey>
+            : ISpanSortHelper<TKey>
+            where TKey : IComparable<TKey>
+        {
+            public void Sort(ref TKey keys, int length)
+            {
+                S.Sort(ref keys, length);
+            }
+
+            public void Sort(ref TKey keys, int length, Comparison<TKey> comparison)
+            {
+                // TODO: Check if comparison is Comparer<TKey>.Default.Compare
+
+                S.Sort(ref keys, length, comparison);
+            }
+        }
+
+
+        internal static class DefaultSpanSortHelper<TKey, TComparer>
+            where TComparer : IComparer<TKey>
+        {
+            internal static readonly ISpanSortHelper<TKey, TComparer> s_default = CreateSortHelper();
+
+            private static ISpanSortHelper<TKey, TComparer> CreateSortHelper()
+            {
+                if (typeof(IComparable<TKey>).IsAssignableFrom(typeof(TKey)))
+                {
+                    // coreclr uses RuntimeTypeHandle.Allocate
+                    var ctor = typeof(ComparableSpanSortHelper<,>)
+                        .MakeGenericType(new Type[] { typeof(TKey), typeof(TComparer) })
+                        .GetConstructor(Array.Empty<Type>());
+
+                    return (ISpanSortHelper<TKey, TComparer>)ctor.Invoke(Array.Empty<object>());
+                }
+                else
+                {
+                    return new SpanSortHelper<TKey, TComparer>();
+                }
+            }
+        }
+
+        // https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Collections/Generic/ArraySortHelper.cs
+        internal interface ISpanSortHelper<TKey, TComparer>
+            where TComparer : IComparer<TKey>
+        {
+            void Sort(ref TKey keys, int length, TComparer comparer);
+        }
+
+        internal class SpanSortHelper<TKey, TComparer> : ISpanSortHelper<TKey, TComparer>
+            where TComparer : IComparer<TKey>
+        {
+            public void Sort(ref TKey keys, int length, TComparer comparer)
+            {
+                // Add a try block here to detect IComparers (or their
+                // underlying IComparables, etc) that are bogus.
+                //
+                // TODO: Do we need the try/catch?
```

Comment:
Let's keep the implementation as close as ArraySortHelper unless there is a known reason to change it.

Reply:
Range check has been implemented by proper checks in `PickPivotAndPartition`, tests verify this and assert that a faulted sort still gives the same output as Array.Sort.